narwhals/_spark_like/expr_str.py

```diff
@@ -107,47 +108,28 @@ def to_lowercase(self: Self) -> SparkLikeExpr:
         )
 
     def to_datetime(self: Self, format: str | None) -> SparkLikeExpr:  # noqa: A002
-        is_naive = (
-            format is not None
-            and "%s" not in format
-            and "%z" not in format
-            and "Z" not in format
-        )
-        function = (
-            self._compliant_expr._F.to_timestamp_ntz
-            if is_naive
-            else self._compliant_expr._F.to_timestamp
-        )
-        pyspark_format = strptime_to_pyspark_format(format)
-        format = (
-            self._compliant_expr._F.lit(pyspark_format) if is_naive else pyspark_format
-        )
+        F = self._compliant_expr._F  # noqa: N806
+        if not format:
+            function = F.to_timestamp
+        elif is_naive(format):
+            function = partial(
+                F.to_timestamp_ntz, format=F.lit(strptime_to_pyspark_format(format))
+            )
+        else:
+            function = partial(F.to_timestamp, format=strptime_to_pyspark_format(format))
         return self._compliant_expr._from_call(
-            lambda _input: function(
-                self._compliant_expr._F.replace(
-                    _input,
-                    self._compliant_expr._F.lit("T"),
-                    self._compliant_expr._F.lit(" "),
-                ),
-                format=format,
-            ),
+            lambda _input: function(F.replace(_input, F.lit("T"), F.lit(" "))),
             "to_datetime",
         )
 
 
-@overload
-def strptime_to_pyspark_format(format: None) -> None: ...
-
-
-@overload
-def strptime_to_pyspark_format(format: str) -> str: ...
+def is_naive(obj: Any) -> bool:
+    return obj is not None and {"%s", "%z", "Z"}.isdisjoint(obj)
```

Comment:
```suggestion
def is_naive(format: str) -> bool:
    return {"%s", "%z", "Z"}.isdisjoint(format)
```

just to signal that this checks the date time format (should we maybe call it `is_naive_format`?). 
Also based on how it is used, `format` will always be a non-null str right?

Reply:
Thanks @EdAbati, yeah `is_naive_format` sounds good to me

I renamed `format` to `obj` to avoid a ruff warning for shadowing `builtins.format`

Also narrowing from `Any` to `str` sounds good as well.
I originally wrote it using a `TypeIs` but removed that since it gave a false-positive on unreachability

Reply:
Happy for you to make those changes and merge 🙌

Reply:
Actually I think the current `set` needs to be changed to:

```py
{"s", "z", "Z"}
```

I'm relying on `format` splitting into a sequence of characters - but two of them a 2-char sequences.

I wonder why that didn't fail a test?

Reply:
> I wonder why that didn't fail a test?

I actually haven't seen a test covering this 👀 double checking with the datetime wizard @MarcoGorelli , is it something that we are not testing on purpose?

Reply:
> > I wonder why that didn't fail a test?
> 
> I actually haven't seen a test covering this 👀 double checking with the datetime wizard @MarcoGorelli , is it something that we are not testing on purpose?

@EdAbati 😮 

IMO there's more than enough `narwhals`-own logic here to warrant testing

Reply:
Probably related to
- #1818

Reply:
Maybe I'm reading this wrong, but I dont think we have coverage for **any backend** with a tz-aware format for `.str.to_datetime(...)`

https://github.com/narwhals-dev/narwhals/blob/3103b20afc1a523137f74003478c77a9debc2781/tests/expr_and_series/str/to_datetime_test.py

Reply:
@MarcoGorelli really need some help on writing a test for this 🙏 

Right now I've got something *kinda* working for a lot of backends.
But it fails to parse for `sqlframe`, `pyarrow`.

Also [`polars.Expr.str.to_datetime`](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.str.to_datetime.html) seems to have more options - that might be able to preserve the timezone offset?

### `test_to_datetime_tz_aware`

```py
@pytest.mark.parametrize(
    ("fmt", "data", "expected_pandas", "expected_polars_duckdb"),
    [
        (
            "%Y-%m-%d %H:%M:%S%z",
            {"a": ["2020-01-01 12:34:56+0200"]},
            "2020-01-01 12:34:56+02:00",
            "2020-01-01 10:34:56+00:00",
        )
    ],
)
def test_to_datetime_tz_aware(
    constructor: Constructor,
    fmt: str,
    data: dict[str, Sequence[str]],
    expected_pandas: str,
    expected_polars_duckdb: str,
) -> None:
    constructor_str = str(constructor)
    if "pandas" in constructor_str or "dask" in constructor_str:
        expected = expected_pandas
    elif "polars" in constructor_str or "duckdb" in constructor_str:
        expected = expected_polars_duckdb
    else:
        expected = "fails"
    result = (
        nw.from_native(constructor(data))
        .lazy()
        .select(b=nw.col("a").str.to_datetime(fmt))
        .collect()
        .item(row=0, column="b")
    )

    assert str(result) == expected
```

I can fully understand nope-ing out of handling timezones for all these different packages 😒

Reply:
Alright so (https://github.com/narwhals-dev/narwhals/pull/2152/commits/e19dcdf5f747a350c69b2a06236bbdcd886d6b27) -> (https://github.com/narwhals-dev/narwhals/pull/2152/commits/194ef83ca93d9c91c9efd3626251643e42f458bd) now has a single test

`sqlframe` works, `pyspark` apprently doesn't https://github.com/narwhals-dev/narwhals/actions/runs/13700361438/job/38312364899?pr=2152

Absolutely don't want to do timezones again 😩

Reply:
@FBruzzesi I see you added `strp_to_pyspark_format` in #1826

AFAICT there were never any tests for tz-aware timestamps and right now I can't get one to work 😞 

Tried my best, but as of (https://github.com/narwhals-dev/narwhals/pull/2152/commits/ce3145db05be055d3387260351d12538b5b3dc49) I'm giving up